### PR TITLE
Feature/cloud 1484 sns module

### PR DESCRIPTION
**Pull request by naveen-variant**
# Description

Terraform module to create sns topic.

Fixes [#2 ](https://usxtech.atlassian.net/browse/CLOUD-1484)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

```yaml
Pre reqs:
  - Clone/Pull branch
  - Create terraform.tfvars.json 
{
    "aws_resource_name_prefix":"dpl-",
    "name": "test-topic-naveen",
    "kms_key_sns_alias_arn":"arn:aws:kms:us-east-1:786352483360:key/255fb0cb-ac0b-4237-acf8-93ffd62c988e"
}

  - Create provider.tf (provided in readme.md)

Test Case 1 [Create SNS topic]:
- Steps
  - Terraform init
  - Terraform plan/apply
- Result
 - Should create sns topic

Test Case 2 [Destroy Sns topic]:
- Steps
  - terraform destroy
- Results
  - See Bucket deleted
```
- [x] Sanity Testing

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added documentation to test

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  create_sns_topic                 = true
+  name                             = "${var.aws_resource_name_prefix}${var.name}"
+  display_name                     = var.display_name
+  fifo_topic                       = var.fifo_topic
+  content_based_deduplication      = var.content_based_deduplication
+  delivery_policy                  = var.delivery_policy
+  sqs_success_feedback_role_arn    = var.sqs_success_feedback_role_arn
+  sqs_success_feedback_sample_rate = var.sqs_success_feedback_sample_rate
+  sqs_failure_feedback_role_arn    = var.sqs_failure_feedback_role_arn
+  kms_master_key_id                = data.aws_kms_key.sns_alias.arn
```
Comment (salyateem13) [inline main.tf]: kms key should be a variable

Comment (naveen-variant) [reply main.tf]: I thought we decided again to add it here. Anyways i can change it.

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  content_based_deduplication      = var.content_based_deduplication
+  delivery_policy                  = var.delivery_policy
+  sqs_success_feedback_role_arn    = var.sqs_success_feedback_role_arn
+  sqs_success_feedback_sample_rate = var.sqs_success_feedback_sample_rate
+  sqs_failure_feedback_role_arn    = var.sqs_failure_feedback_role_arn
+  kms_master_key_id                = var.kms_key_sns_alias_arn
+}
+
+# tflint-ignore: terraform_unused_declarations
+data "aws_iam_policy_document" "sns_publish_policy" {
```
Comment (LeeHughsVariant) [inline main.tf:18]: What is the point of this data block? I don't see it being used for anything

Comment (naveen-variant) [reply main.tf:18]: Corrected to add as output.

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,59 @@
+variable "aws_resource_name_prefix" {
+  type        = string
+  description = "Prefix of team name to be applied to created resources."
+}
+
```
Comment (salyateem13) [inline variables.tf:5]: extra line

Comment (salyateem13) [inline variables.tf:5]: extra line

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  create_sns_topic                 = true
+  name                             = "${var.aws_resource_name_prefix}${var.name}"
+  display_name                     = var.display_name
+  fifo_topic                       = var.fifo_topic
+  content_based_deduplication      = var.content_based_deduplication
+  delivery_policy                  = var.delivery_policy
+  sqs_success_feedback_role_arn    = var.sqs_success_feedback_role_arn
+  sqs_success_feedback_sample_rate = var.sqs_success_feedback_sample_rate
+  sqs_failure_feedback_role_arn    = var.sqs_failure_feedback_role_arn
+  kms_master_key_id                = var.kms_key_sns_alias_arn
```
Comment (salyateem13) [inline main.tf:13]: We should add tags as input variable

Comment (salyateem13) [inline main.tf:13]: We should add tags as input variable

File: versions.tf [hunk 5]
```diff
@@ -0,0 +1,8 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 3.74"
```
Comment (salyateem13) [inline versions.tf]: Test with ~> 4.4.0 (latest)

File: variables.tf (hunk 3 above)
Comment (naveen-variant) [reply variables.tf:5]: Corrected

Comment (naveen-variant) [reply variables.tf:5]: Corrected

File: versions.tf (hunk 5 above)
Comment (naveen-variant) [reply versions.tf]: Added

File: main.tf (hunk 4 above)
Comment (naveen-variant) [reply main.tf:13]: Added

Comment (naveen-variant) [reply main.tf:13]: Added

**Comment by salyateem13** [comment]
Same here, add example please

**Comment by rchengvariant** [comment]
@naveen-variant Same comment as Samir, please create examples.
```
Test Case 1 [Create SNS topic]:
- Steps
  - Terraform init
  - Terraform plan/apply
- Result
 - Should create sns topic
 - PASSED

Test Case 2 [Destroy Sns topic]:
- Steps
  - terraform destroy
- Results
  - See SNS deleted
  - PASSED
```

**Comment by naveen-variant** [comment]
@salyateem13 @rchengvariant  Added examples.

**Review by rchengvariant — APPROVED** [review]

**Review by salyateem13 — APPROVED** [review]

**Pull request merged**